File: app/infrastructure/qaac_runner.py
# -*- coding: utf-8 -*-
"""Модуль для запуска кодировщика QAAC."""

import logging
import subprocess
import os
from pathlib import Path

from app.core import path_utils
from app.core.singleton import SingletonMeta
from app.core.process_manager import ProcessManager

logger = logging.getLogger(__name__)


class QaacRunner(metaclass=SingletonMeta):
    """Класс для управления процессом qaac64.exe."""

    def __init__(self) -> None:
        """Инициализация runner'а."""
        self.__qaac_path: str | None = None
        self.__ffmpeg_path: str | None = None

    @property
    def _qaac_path(self) -> str:
        if self.__qaac_path is None:
            self.__qaac_path = path_utils.get_binary_path("qaac64")
            logger.debug(
                "QaacRunner инициализирован. qaac: %s", self.__qaac_path
            )
        return self.__qaac_path

    @property
    def _ffmpeg_path(self) -> str:
        if self.__ffmpeg_path is None:
            self.__ffmpeg_path = path_utils.get_binary_path("ffmpeg")
        return self.__ffmpeg_path

    def run(
        self,
        input_path: Path,
        output_path: Path,
        tvbr: str = "127",
        adts: bool = False,
        extra_args: list[str] | None = None,
        overwrite: bool = False,
    ) -> bool:
        """Запустить qaac64 через конвейер с FFmpeg.

        Это позволяет поддерживать любые входные форматы FFmpeg.

        Args:
            input_path: Путь к входному файлу.
            output_path: Путь к выходному файлу.
            tvbr: Уровень качества.
            adts: Формат ADTS.
            extra_args: Доп. аргументы.
            overwrite: Флаг перезаписи (переопределяется).

        Returns:
            True при успехе, False при ошибке.
        """
        if not Path(self._qaac_path).exists():
            logger.error("Бинарник qaac64.exe не найден: %s", self._qaac_path)
            return False

        ffmpeg_cmd = self._build_ffmpeg_cmd(input_path)
        qaac_cmd = self._build_qaac_cmd(output_path, tvbr, adts, extra_args)
        env = self._prepare_env()

        logger.info(
            "Запуск конвейера: %s | %s",
            " ".join(ffmpeg_cmd),
            " ".join(qaac_cmd),
        )

        return self._execute_pipeline(
            ffmpeg_cmd, qaac_cmd, env, input_path.name
        )

    def _build_ffmpeg_cmd(self, input_path: Path) -> list[str]:
        """Сформировать аргументы декодера ffmpeg."""
        return [
            self._ffmpeg_path,
            "-v",
            "error",
            "-i",
            str(input_path),
            "-f",
            "wav",
            "-",
        ]

    def _build_qaac_cmd(
        self,
        output_path: Path,
        tvbr: str,
        adts: bool,
        extra_args: list[str] | None,
    ) -> list[str]:
        """Сформировать аргументы кодировщика qaac."""
        cmd = [
            self._qaac_path,
            "--tvbr",
            tvbr,
            "-",
            "-o",
            str(output_path),
        ]
        if adts:
            cmd.insert(1, "--adts")
        if extra_args:
            cmd.extend(extra_args)
        return cmd

    def _prepare_env(self) -> dict[str, str]:
        """Настройка окружения для Apple Application Support."""
        base_dir = Path(self._qaac_path).parent
        pf = os.path.expandvars(
            r"%ProgramFiles%\Common Files\Apple\Apple Application Support"
        )
        pfx86 = os.path.expandvars(
            r"%ProgramFiles(x86)%\Common Files\Apple\Apple Application Support"
        )

        apple_paths = [
            str(base_dir / "QTFiles64"),
            str(base_dir / "QTFiles"),
            pf,
            pfx86,
        ]

        env = os.environ.copy()
        valid_paths = [str(base_dir)]

        for p in apple_paths:
            if os.path.isdir(p):
                valid_paths.append(p)

        env["PATH"] = (
            os.pathsep.join(valid_paths) + os.pathsep + env.get("PATH", "")
        )
        return env

    def _execute_pipeline(
        self,
        ffmpeg_cmd: list[str],
        qaac_cmd: list[str],
        env: dict[str, str],
        input_name: str,
    ) -> bool:
        """Запуск конвейера процессов и обработка ошибок."""
        try:
            ffmpeg_proc = subprocess.Popen(
                ffmpeg_cmd,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                creationflags=subprocess.CREATE_NO_WINDOW,
            )
            ProcessManager().register(ffmpeg_proc)

            qaac_proc = subprocess.Popen(
                qaac_cmd,
                stdin=ffmpeg_proc.stdout,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
                encoding="utf-8",
                errors="replace",
                env=env,
                creationflags=subprocess.CREATE_NO_WINDOW,
            )
            ProcessManager().register(qaac_proc)

            if ffmpeg_proc.stdout:
                ffmpeg_proc.stdout.close()

            try:
                qaac_stdout, qaac_stderr = qaac_proc.communicate()
                ffmpeg_stderr = (
                    ffmpeg_proc.stderr.read().decode("utf-8", "replace")
                    if ffmpeg_proc.stderr
                    else ""
                )

                ffmpeg_proc.wait()
            finally:
                ProcessManager().unregister(qaac_proc)
                ProcessManager().unregister(ffmpeg_proc)

            return self._check_pipeline_results(
                qaac_proc, ffmpeg_proc, qaac_stderr, ffmpeg_stderr
            )

        except Exception:
            logger.exception(
                "Критическая ошибка конвейера QAAC для '%s'", input_name
            )
            return False

    def _check_pipeline_results(
        self,
        qaac_proc: subprocess.Popen,
        ffmpeg_proc: subprocess.Popen,
        qaac_stderr: str,
        ffmpeg_stderr: str,
    ) -> bool:
        """Проверка кодов возврата процессов конвейера."""
        from app.core.process_manager import ProcessManager

        if ProcessManager().was_cancelled(
            qaac_proc
        ) or ProcessManager().was_cancelled(ffmpeg_proc):
            logger.info("Процесс QAAC был прерван пользователем.")
            return False

        if qaac_proc.returncode != 0:
            logger.error(
                "QAAC завершился с ошибкой (%d).\nSTDERR: %s\nFFmpeg STDERR: %s",  # noqa: E501
                qaac_proc.returncode,
                qaac_stderr.strip(),
                ffmpeg_stderr.strip(),
            )
            return False

        if ffmpeg_proc.returncode != 0:
            logger.error(
                "FFmpeg (декодер) завершился с ошибкой (%d).\nSTDERR: %s",
                ffmpeg_proc.returncode,
                ffmpeg_stderr.strip(),
            )
            return False

        return True

File: src-csharp/KTools.App/Infrastructure/QaacRunner.cs
// -*- coding: utf-8 -*-
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using KTools_App.Core;

namespace KTools_App.Infrastructure;

/// <summary>
/// Синглтон-обертка для запуска кодировщика Apple AAC (qaac64.exe) через конвейер с FFmpeg.
/// Позволяет кодировать любые медиаформаты, поддерживаемые FFmpeg, напрямую в высококачественный AAC/M4A 
/// без создания промежуточных временных WAV-файлов на диске.
/// Все комментарии и логирование выполнены строго на русском языке в соответствии с регламентом.
/// </summary>
public sealed class QaacRunner
{
    private static readonly Lazy<QaacRunner> LazyInstance =
        new(() => new QaacRunner());

    private QaacRunner() { }

    /// <summary>
    /// Возвращает единственный экземпляр класса QaacRunner.
    /// </summary>
    public static QaacRunner Instance => LazyInstance.Value;

    /// <summary>
    /// Запустить кодирование AAC через потоковый конвейер FFmpeg | QAAC64.
    /// </summary>
    /// <param name="inputPath">Абсолютный путь к исходному медиафайлу.</param>
    /// <param name="outputPath">Абсолютный путь к выходному файлу M4A/AAC.</param>
    /// <param name="tvbr">Уровень качества переменного битрейта (True VBR, например, "127").</param>
    /// <param name="adts">Использовать ли контейнер ADTS (расширение .aac) вместо M4A.</param>
    /// <param name="extraArgs">Дополнительные параметры для qaac64.</param>
    /// <param name="cancellationToken">Токен отмены задачи.</param>
    /// <returns>True в случае успеха, иначе false.</returns>
    public async Task<bool> RunAsync(
        string inputPath,
        string outputPath,
        string tvbr = "127",
        bool adts = false,
        List<string>? extraArgs = null,
        CancellationToken cancellationToken = default)
    {
        string qaacPath = PathManager.GetBinaryPath("qaac64");
        string ffmpegPath = PathManager.GetBinaryPath("ffmpeg");

        if (!File.Exists(qaacPath))
        {
            LogService.Instance.Error($"Критическая ошибка: отсутствует кодировщик qaac64.exe по пути: '{qaacPath}'", "QaacRunner");
            return false;
        }

        if (!File.Exists(ffmpegPath))
        {
            LogService.Instance.Error($"Критическая ошибка: отсутствует декодер ffmpeg.exe по пути: '{ffmpegPath}'", "QaacRunner");
            return false;
        }

        LogService.Instance.Info($"Начало кодирования QAAC (TVBR: {tvbr}) для файла: '{Path.GetFileName(inputPath)}'", "QaacRunner");

        // 1. Формируем аргументы для FFmpeg (декодирование в WAV и вывод в stdout)
        string ffmpegArgs = $"-v error -i \"{inputPath}\" -f wav -";

        // 2. Формируем аргументы для QAAC (чтение из stdin и запись в файл)
        var qaacArgsList = new List<string>();
        if (adts)
        {
            qaacArgsList.Add("--adts");
        }
        qaacArgsList.Add("--tvbr");
        qaacArgsList.Add(tvbr);
        qaacArgsList.Add("-"); // Вход из stdin
        qaacArgsList.Add("-o");
        qaacArgsList.Add($"\"{outputPath}\"");

        if (extraArgs != null)
        {
            qaacArgsList.AddRange(extraArgs);
        }

        string qaacArgs = string.Join(" ", qaacArgsList);

        // 3. Подготовка переменных окружения Apple Application Support
        var env = PrepareAppleEnvironment(qaacPath);

        LogService.Instance.DebugLog($"Запуск конвейера: ffmpeg {ffmpegArgs} | qaac64 {qaacArgs}", "QaacRunner");

        // 4. Настройка процессов
        var ffmpegStartInfo = new ProcessStartInfo
        {
            FileName = ffmpegPath,
            Arguments = ffmpegArgs,
            CreateNoWindow = true,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            WorkingDirectory = Path.GetDirectoryName(ffmpegPath) ?? AppContext.BaseDirectory
        };

        var qaacStartInfo = new ProcessStartInfo
        {
            FileName = qaacPath,
            Arguments = qaacArgs,
            CreateNoWindow = true,
            UseShellExecute = false,
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            WorkingDirectory = Path.GetDirectoryName(qaacPath) ?? AppContext.BaseDirectory
        };

        // Заполняем переменные окружения Apple
        foreach (var pair in env)
        {
            qaacStartInfo.EnvironmentVariables[pair.Key] = pair.Value;
        }

        using var ffmpegProc = new Process { StartInfo = ffmpegStartInfo };
        using var qaacProc = new Process { StartInfo = qaacStartInfo };

        try
        {
            ffmpegProc.Start();
            qaacProc.Start();
            
            LogService.Instance.DebugLog($"Запущены процессы конвейера. FFmpeg PID: {ffmpegProc.Id}, QAAC PID: {qaacProc.Id}", "QaacRunner");
        }
        catch (Exception ex)
        {
            LogService.Instance.Exception(ex, $"Не удалось запустить процессы конвейера QAAC: {ex.Message}", "QaacRunner");
            try { ffmpegProc.Kill(true); } catch { }
            try { qaacProc.Kill(true); } catch { }
            return false;
        }

        // Задачи для перенаправления вывода ошибок
        var ffmpegStderrLines = new List<string>();
        var qaacStderrLines = new List<string>();

        var ffmpegErrorTask = Task.Run(async () =>
        {
            while (!ffmpegProc.StandardError.EndOfStream)
            {
                string? line = await ffmpegProc.StandardError.ReadLineAsync();
                if (line != null)
                {
                    lock (ffmpegStderrLines)
                    {
                        ffmpegStderrLines.Add(line);
                    }
                }
            }
        });

        var qaacErrorTask = Task.Run(async () =>
        {
            while (!qaacProc.StandardError.EndOfStream)
            {
                string? line = await qaacProc.StandardError.ReadLineAsync();
                if (line != null)
                {
                    lock (qaacStderrLines)
                    {
                        qaacStderrLines.Add(line);
                    }
                }
            }
        });

        // 5. Конвейеризация данных из stdout FFmpeg в stdin QAAC
        var pipeTask = Task.Run(async () =>
        {
            try
            {
                byte[] buffer = new byte[65536]; // 64KB буфер
                using var input = ffmpegProc.StandardOutput.BaseStream;
                using var output = qaacProc.StandardInput.BaseStream;

                int bytesRead;
                while ((bytesRead = await input.ReadAsync(buffer, 0, buffer.Length, cancellationToken)) > 0)
                {
                    await output.WriteAsync(buffer, 0, bytesRead, cancellationToken);
                }
            }
            catch (OperationCanceledException)
            {
                // Игнорируем при штатной отмене
            }
            catch (Exception ex)
            {
                LogService.Instance.Error($"Ошибка конвейерной передачи данных FFmpeg -> QAAC: {ex.Message}", "QaacRunner");
            }
            finally
            {
                try
                {
                    qaacProc.StandardInput.BaseStream.Close(); // Закрываем поток, чтобы qaac понял конец файла
                }
                catch { }
            }
        });

        // Ожидание завершения процессов или отмены
        try
        {
            var waitFfmpeg = ffmpegProc.WaitForExitAsync(cancellationToken);
            var waitQaac = qaacProc.WaitForExitAsync(cancellationToken);

            await Task.WhenAll(waitFfmpeg, waitQaac, pipeTask, ffmpegErrorTask, qaacErrorTask);
        }
        catch (OperationCanceledException)
        {
            LogService.Instance.Warn("Конвейер QAAC прерван пользователем. Принудительная остановка процессов...", "QaacRunner");
            try { ffmpegProc.Kill(true); } catch { }
            try { qaacProc.Kill(true); } catch { }
            return false;
        }

        // Проверка результатов
        bool success = ffmpegProc.ExitCode == 0 && qaacProc.ExitCode == 0;

        if (!success)
        {
            string ffErr = string.Join(Environment.NewLine, ffmpegStderrLines);
            string qaacErr = string.Join(Environment.NewLine, qaacStderrLines);
            LogService.Instance.Error($"Сбой конвейера QAAC.\nFFmpeg Code: {ffmpegProc.ExitCode}, Err: {ffErr}\nQAAC Code: {qaacProc.ExitCode}, Err: {qaacErr}", "QaacRunner");
            return false;
        }

        LogService.Instance.Info($"Конвейер QAAC успешно завершил работу: '{Path.GetFileName(outputPath)}'", "QaacRunner");
        return true;
    }

    /// <summary>
    /// Настраивает PATH для Apple Application Support, необходимый для работы библиотек QAAC.
    /// </summary>
    private Dictionary<string, string> PrepareAppleEnvironment(string qaacPath)
    {
        string baseDir = Path.GetDirectoryName(qaacPath) ?? AppContext.BaseDirectory;
        
        // Стандартные пути Apple Application Support в системе
        string pf = Environment.ExpandEnvironmentVariables(@"%ProgramFiles%\Common Files\Apple\Apple Application Support");
        string pfx86 = Environment.ExpandEnvironmentVariables(@"%ProgramFiles(x86)%\Common Files\Apple\Apple Application Support");

        var applePaths = new List<string>
        {
            baseDir,
            Path.Combine(baseDir, "QTFiles64"),
            Path.Combine(baseDir, "QTFiles")
        };

        if (Directory.Exists(pf)) applePaths.Add(pf);
        if (Directory.Exists(pfx86)) applePaths.Add(pfx86);

        // Получаем текущую переменную PATH
        string currentPath = Environment.GetEnvironmentVariable("PATH") ?? string.Empty;
        
        // Объединяем пути через разделитель
        string newPath = string.Join(Path.PathSeparator, applePaths) + Path.PathSeparator + currentPath;

        var env = new Dictionary<string, string>();
        env["PATH"] = newPath;
        return env;
    }
}
